File: ballast-client/npm/src/signalr/signalr-client-service-base.ts
import * as signalR from '@aspnet/signalr';
import { Guid, IDisposable, IEventBus, IEvent } from 'ballast-core';
import { injectable } from 'inversify';
import { ISignalRClientServiceOptions } from './signalr-client-service-options';

type InvocationResolver<TValue extends any> = (value?: TValue) => void;
type InvocationRejector = (reason?: string) => void;

@injectable()
export abstract class SignalRClientServiceBase implements IDisposable {

    protected readonly _optionsFactory: () => ISignalRClientServiceOptions;
    protected readonly _methods: Map<string, string>;
    protected readonly _subscriptions: Map<string, string>;
    protected readonly _invocations: Map<string, Map<string, any>>;
    protected _hubConnection?: signalR.HubConnection;
    protected _hubConnectionState: 'disconnected' | 'connected' | 'connecting';

    protected abstract get hubName(): string;

    protected afterSubscribe(hubConnection: signalR.HubConnection): void { }
    protected beforeUnsubscribe(hubConnection: signalR.HubConnection): void { }
    protected onDispose() { }

    public constructor(optionsFactory: () => ISignalRClientServiceOptions) {
        this._optionsFactory = optionsFactory;
        this._methods = new Map<string, string>();
        this._subscriptions = new Map<string, string>();
        this._invocations = new Map<string, Map<string, any>>();
        this._hubConnectionState = 'disconnected';
    }

    public dispose() {
        if (this._hubConnection) {
            this.disconnectAsync(); // Fire and forget
        }
        this.onDispose();
    }

    public get isConnected() {
        return (!!this._hubConnection && this._hubConnectionState == 'connected');
    }

    public get isConnecting() {
        return (!!this._hubConnection && this._hubConnectionState == 'connecting');
    }

    public createHubConnection() {
        let options = this._optionsFactory();
        let hubName = this.hubName;
        let hub = `${options.serverUrl}/${hubName}`;
        let connectionBuilder = new signalR.HubConnectionBuilder();
        let connection = connectionBuilder.withUrl(hub).build();
        return connection;
    }

    public async connectAsync() {
        if (this.isConnected) {
            await this.disconnectAsync();
        }
        this._hubConnectionState = 'connecting';
        this._hubConnection = this.createHubConnection();
        this.resubscribeToHubEvents();
        await this._hubConnection.start();
        this._hubConnectionState = 'connected';
        await this.registerClientAsync();
    }

    private async registerClientAsync() {
        let options = this._optionsFactory();
        let clientId = options.clientId;
        await this.createInvocationAsync('registerClientAsync', clientId);
    }

    public async disconnectAsync() {
        if (this._hubConnection) {
            this.unsubscribeFromHubEvents();
            await this._hubConnection.stop();
        }
        this._hubConnection = undefined;
        this._hubConnectionState = 'disconnected';
    }

    protected registerHubMethod(method: string) {
        this._methods.set(method, method);
        this.registerCallbackForHubMethod(method);
    }

    private registerCallbackForHubMethod(methodName: string) {
        // Do not proceed unless are have a hub connection
        if (!this._hubConnection) {
            return;
        }
        // Register fulfillment (callback) subscription for the current method
        let callback = `${methodName}Callback`;
        this._subscriptions.set(callback, callback);
        this._hubConnection.on(callback, (invocationId: string, reason?: null | string, value?: null | any) => {
            // Get the list of all currently running/live method invocations
            let currentInvocations = this.getInvocationList(methodName);
            let foundInvocation = currentInvocations.get(invocationId);
            if (!foundInvocation) {
                return;
            }
            let resolve = foundInvocation["0"] as InvocationResolver<any>;
            let reject = foundInvocation["1"] as InvocationRejector;
            if (!!reason) {
                // Reject the promise
                reject(reason);
            } else {
                // Resolve the promise
                resolve(value || undefined);
            }
            // Remove from the list of invocations (promise has reached fulfilled state)
            currentInvocations.delete(invocationId);
        });
    }

    protected onConnectionClosed() {
        // Try to re-open
        this._hubConnectionState = 'disconnected';
        if (this._hubConnection) {
            this._hubConnectionState = 'connecting';
            this._hubConnection.start()
                .then(() => this._hubConnectionState = 'connected');
        }
    }

    protected resubscribeToHubEvents() {
        // Unsubscribe from all (if already subscribed)
        this.unsubscribeFromHubEvents();
        // Make sure we have a hub connection instance
        if (!this._hubConnection) {
            throw new Error('Cannot (re)subscribe to hub events without a hub connection');
        }
        // Iterate through registered hub methods
        this._methods.forEach(methodName => this.registerCallbackForHubMethod(methodName));
        this.afterSubscribe(this._hubConnection);
        this._hubConnection.onclose = this.onConnectionClosed.bind(this);
    }

    protected unsubscribeFromHubEvents() {
        if (this._hubConnection) {
            this.beforeUnsubscribe(this._hubConnection);
            let subscriptions = Array.from(this._subscriptions.keys());
            for (let subscription of subscriptions) {
                this._hubConnection.off(subscription);
            }
        }
        this._subscriptions.clear();
    }

    protected createInvocationId() {
        return Guid.newGuid();
    }

    protected getInvocationList(method: string) {
        if (!this._invocations.has(method)) {
            this._invocations.set(method, new Map());
        }
        return this._invocations.get(method)!;
    }

    protected async createInvocationAsync<TValue extends any>(method: string, ...args: any[]) {
        if (!this._methods.has(method)) {
            this.registerHubMethod(method);
        }
        if (this.isConnecting) {
            throw new Error('Invocation could not be created because the service is still attempting to establish a connection')
        }
        if (!this.isConnected) {
            await this.connectAsync();
        }
        return await new Promise<TValue>((resolve, reject) => {
            let invocationList = this.getInvocationList(method);
            let invocationId = this.createInvocationId();
            invocationList.set(invocationId, [resolve, reject]);
            this.invokeOnHubAsync(method, invocationId, ...args)
                .catch((err: Error) => this.cancelInvocation(method, invocationId, err.message)); // Fire and forget
        });
    }

    private cancelInvocation(method: string, invocationId: string, cancellationReason: string) {
        // Get the list of all currently running/live method invocations
        let currentInvocations = this.getInvocationList(method);
        let foundInvocation = currentInvocations.get(invocationId);
        if (!foundInvocation) {
            return;
        }
        let reject = foundInvocation["1"] as InvocationRejector;
        // Reject the promise
        reject(cancellationReason);
        // Remove from the list of invocations (promise has reached fulfilled state)
        currentInvocations.delete(invocationId);
    }

    private async invokeOnHubAsync(method: string, invocationId: string, ...args: any[]) {
        if (this.isConnecting) {
            throw new Error('Invocation could not take place because the service is still attempting to establish a connection')
        }
        if (!this.isConnected) {
            await this.connectAsync()
        } 
        if (this.isConnected) {
            let invocationIdPlusArgs = (<any[]>[invocationId]).concat(args);
            await this._hubConnection!.invoke(method, ...invocationIdPlusArgs);
        }
    }

}

File: ballast-client/dotnet/src/signalr/SignalRClientServiceBase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;

namespace Ballast.Client.SignalR.Services
{
    public abstract class SignalRClientServiceBase : IDisposable
    {

        private readonly ISignalRClientOptions _options;
        private readonly HashSet<string> _methods;
        private readonly HashSet<string> _subscriptions;
        private readonly IDictionary<string, IDictionary<Guid, TaskCompletionSource<object>>> _invocations;
        private HubConnection _hubConnection;
        private string _hubConnectionState;

        private const string STATE_DISCONNECTED = "disconnected";
        private const string STATE_CONNECTED = "connected";
        private const string STATE_CONNECTING = "connecting";

        protected abstract string HubName { get; }

        protected virtual void AfterSubscribe(HubConnection hubConnection) { }
        protected virtual void BeforeUnsubscribe(HubConnection hubConnection) { }
        protected virtual void OnDispose() { }

        public SignalRClientServiceBase(ISignalRClientOptions options) 
        {
            _options = options;
            _methods = new HashSet<string>();
            _subscriptions = new HashSet<string>();
            _invocations = new Dictionary<string, IDictionary<Guid, TaskCompletionSource<object>>>();
            _hubConnectionState = STATE_DISCONNECTED;
        }

        public void Dispose()
        {
            if (_hubConnection != null) 
                _ = DisconnectAsync(); // Fire and forget
            OnDispose();
        }

        public bool IsConnected => _hubConnection != null && _hubConnectionState == STATE_CONNECTED;

        public bool IsConnecting => _hubConnection != null && _hubConnectionState == STATE_CONNECTING;

        private HubConnection CreateHubConnection()
        {
            var options = _options;
            var hubName = HubName;
            var hub = $"{options.ServerUrl}/{hubName}";
            var connectionBuilder = new HubConnectionBuilder();
            var connection = connectionBuilder.WithUrl(hub).Build();
            return connection;
        }

        public async Task ConnectAsync()
        {
            if (IsConnected)
                await DisconnectAsync();
            _hubConnectionState = STATE_CONNECTING;
            _hubConnection = CreateHubConnection();
            ResubscribeToHubEvents();
            await _hubConnection.StartAsync();
            _hubConnectionState = STATE_CONNECTED;
            await RegisterClientAsync();
        }

        private async Task RegisterClientAsync() 
        {
            var options = _options;
            var clientId = options.ClientId;
            await CreateInvocationAsync<int>("registerClientAsync", clientId);
        }

        public async Task DisconnectAsync()
        {
            if (_hubConnection != null)
            {
                UnsubscribeFromHubEvents();
                await _hubConnection.StopAsync();
            }
            _hubConnection = null;
            _hubConnectionState = STATE_DISCONNECTED;
        }

        protected void RegisterHubMethod(string method)
        {
            _methods.Add(method);
            RegisterCallbackForHubMethod(method);
        }

        private void RegisterCallbackForHubMethod(string methodName)
        {
            // Do not proceed unless are have a hub connection
            if (this._hubConnection == null) {
                return;
            }
            // Register fulfillment (callback) subscription for the current method
            var callback = $"{methodName}Callback";
            _subscriptions.Add(callback);
            _hubConnection.On(callback, (Guid invocationId, string reason, object value) => {
                // Get the list of all currently running/live method invocations
                var currentInvocations = GetInvocationList(methodName);
                currentInvocations.TryGetValue(invocationId, out var foundInvocation);
                if (foundInvocation == null)
                    return;
                if (reason != null)
                    foundInvocation.SetException(new Exception(reason));
                else
                    foundInvocation.SetResult(value); // if there was no error or value object we assume void method (int)
                // Remove from the list of invocations (task fulfilled/completed)
                currentInvocations.Remove(invocationId);
            });
        }

        protected async Task OnConnectionClosed(Exception ex = null)
        {
            // Try to re-open
            _hubConnectionState = STATE_DISCONNECTED;
            if (this._hubConnection != null)
            {
                _hubConnectionState = STATE_CONNECTING;
                await _hubConnection.StartAsync();
                _hubConnectionState = STATE_CONNECTED;
            }
        }

        protected void ResubscribeToHubEvents()
        {
            // Unsubscribe from all (if already subscribed)
            UnsubscribeFromHubEvents();
            // Make sure we have a hub connection instance
            if (_hubConnection == null)
                throw new InvalidOperationException("Cannot (re)subscribe to hub events without a hub connection");
            // Iterate through registered hub methods
            foreach(var methodName in _methods)
            {
                RegisterCallbackForHubMethod(methodName);
            }
            AfterSubscribe(_hubConnection);
            _hubConnection.Closed += OnConnectionClosed;
        }

        protected void UnsubscribeFromHubEvents()
        {
            if (_hubConnection != null)
            {
                BeforeUnsubscribe(_hubConnection);
                var subscriptions = _subscriptions;
                foreach (var subscription in subscriptions)
                {
                    _hubConnection.Remove(subscription);
                }
            }
            _subscriptions.Clear();
        }

        protected Guid CreateInvocationId() => Guid.NewGuid();
        
        protected IDictionary<Guid, TaskCompletionSource<object>> GetInvocationList(string method)
        {
            if (!_invocations.ContainsKey(method))
                _invocations[method] = new Dictionary<Guid, TaskCompletionSource<object>>();
            return _invocations[method];
        }

        protected async Task CreateInvocationAsync(string method, params object[] args)
        {
            _ = await CreateInvocationAsync<object>(method, args);
        }

        protected async Task<TValue> CreateInvocationAsync<TValue>(string method, params object[] args)
        {
            if (!_methods.Contains(method))
                RegisterHubMethod(method);
            if (IsConnecting)
                throw new InvalidOperationException("Invocation could not be created because the service is still attempting to establish a connection");
            if (!IsConnected)
                await ConnectAsync();
            var tcs = new TaskCompletionSource<object>();
            _ = Task.Run(async () =>
            {
                var invocationList = GetInvocationList(method);
                var invocationId = CreateInvocationId();
                invocationList[invocationId] = tcs;
                try
                {
                    await InvokeOnHubAsync(method, invocationId, args);
                }
                catch(Exception ex)
                {
                    tcs.SetException(ex);
                }
            }); // Fire and forget
            // Use Task.ContinueWith to map Task<object> to Task<TValue>
            return await tcs.Task.ContinueWith(x => (TValue)x.Result);
        }

        private void CancelInvocation(string method, Guid invocationId, string cancellationReason)
        {
            // Get the list of all currently running/live method invocations
            var currentInvocations = GetInvocationList(method);
            currentInvocations.TryGetValue(invocationId, out var foundInvocation);
            if (foundInvocation == null)
                return;
            // Cancel/reject the invocation
            foundInvocation.SetException(new InvalidOperationException(cancellationReason));
            // Remove from the list of invocations (fulfilled/completed task)
            currentInvocations.Remove(invocationId);
        }

        private async Task InvokeOnHubAsync(string method, Guid invocationId, params object[] args)
        {
            if (IsConnecting)
                throw new InvalidOperationException("Invocation could not take place because the service is still attempting to establish a connection");
            if (!IsConnected)
                await ConnectAsync();
            if (IsConnected)
            {
                var invocationIdPlusArgs = (new object[] { invocationId }).Concat(args);
                await _hubConnection.InvokeAsync(method, invocationIdPlusArgs);
            }
        }

    }
}